src/module/dy/validate-kwargs.js

```diff
@@ -4,40 +4,64 @@ const schema = Joi.object().keys({
   name: Joi.string(),
   attributes: Joi.object().pattern(
     Joi.string(),
-    // todo: must have exactly one value with partitionKey
-    // todo: must have zero or one value with partitionKey
     Joi.object().keys({
       type: Joi.string().valid('string', 'boolean', 'number', 'list', 'map', 'binary', 'set'),
       partitionKey: Joi.boolean().valid(true).optional(),
       sortKey: Joi.boolean().valid(true).optional()
     }).nand('partitionKey', 'sortKey')
-  ),
+  ).min(1).custom((v, h) => {
+    const propertyCount = {};
+    Object.values(v).forEach((o) => Object.keys(o).forEach((k) => {
+      if (propertyCount[k] === undefined) {
+        propertyCount[k] = 1;
+      } else {
+        propertyCount[k] += 1;
+      }
+    }));
+    if (propertyCount.partitionKey === undefined) {
+      return h.message('At least one partitionKey must be defined');
+    }
+    if (propertyCount.partitionKey > 1) {
+      return h.message('Duplicated partitionKey definition');
+    }
+    if (propertyCount.sortKey !== undefined && propertyCount.sortKey > 1) {
+      return h.message('Duplicated sortKey definition');
+    }
+    return v;
+  }),
   indices: Joi.object().pattern(
     Joi.string(),
     Joi.object().keys({
-      // todo: check that value is present in attributes
       partitionKey: Joi.string().optional(),
-      // todo: check that value is present in attributes
       sortKey: Joi.string().optional()
     }).or('partitionKey', 'sortKey')
-  ),
+  ).optional().min(1)
+    .custom((v, h) => {
+      const valid = Object.values(v)
+        .every((idx) => ([...new Set(Object.values(idx))].length === Object.keys(idx).length));
+      if (valid === false) {
+        return h.message('Can\'t use the same attribute as partitionKey and sortKey');
+      }
+      return v;
+    }),
   DocumentClient: Joi.object()
+}).custom((v, h) => {
+  const { attributes, indices } = v;
+  if (indices !== undefined) {
+    const valid = Object.values(indices)
+      .every((idx) => Object.keys(attributes)
+        .includes(Object.values(idx)[0]));
+    if (valid === false) {
+      return h.message('Indices values must match with defined attributes');
+    }
```

Comment:
I'd first fetch all used keys in the indices and then filter out the ones that are not defined in the attributes and error if that length > 0 (with that array in the details)